porch/engine/pkg/engine/kuberuntime.go

```diff
@@ -0,0 +1,190 @@
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func newKubeFunctionRuntime(cl client.Client, namespace string) (*kubeFunctionRuntime, error) {
+	return &kubeFunctionRuntime{
+		kubeClient: cl,
+		namespace:  namespace,
+	}, nil
+}
+
+type kubeFunctionRuntime struct {
+	// kubeClient is the kubernetes client
+	kubeClient client.Client
+
+	// namespace holds the namespace where the executors run
+	namespace string
+}
+
+var _ kpt.FunctionRuntime = &kubeFunctionRuntime{}
+
+func (kr *kubeFunctionRuntime) GetRunner(ctx context.Context, fn *v1.Function) (fn.FunctionRunner, error) {
+	// Create a pod
+	pod := &corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Pod",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: kr.namespace,
+			Name:      fmt.Sprintf("krm-fn-%v", randStringRunes(6)),
+			// TODO: add labels
+		},
+		Spec: corev1.PodSpec{
+			InitContainers: []corev1.Container{
+				{
+					Name:            "copy-wrapper-server",
+					Image:           "gcr.io/mengqiy-playground/wrapper-server:v0", // TODO
+					ImagePullPolicy: "Always",
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      "wrapper-server",
+							MountPath: "/wrapper-server",
+						},
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Name:    "function",
+					Image:   fn.Image,
+					Command: []string{"/wrapper-server/wrapper-server"},
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      "wrapper-server",
+							MountPath: "/wrapper-server",
+						},
+					},
+				},
+			},
+			Volumes: []corev1.Volume{
+				{
+					Name: "wrapper-server",
+					VolumeSource: corev1.VolumeSource{
+						EmptyDir: &corev1.EmptyDirVolumeSource{},
+					},
+				},
+			},
+		},
+	}
+	err := kr.kubeClient.Create(ctx, pod)
+	if err != nil {
+		return nil, err
+	}
+
+	return &kubeFunctionRunner{
+		ctx:   context.TODO(), // TODO: should we use the provided ctx here?
+		cl:    kr.kubeClient,
+		pod:   pod,
+		image: fn.Image,
+	}, nil
+}
+
+func (k *kubeFunctionRuntime) Close() error {
+	// TODO: cleanup launched pods.
+	return nil
+}
+
+type kubeFunctionRunner struct {
+	ctx   context.Context
+	cl    client.Client
+	pod   *corev1.Pod
+	image string
+}
+
+var _ fn.FunctionRunner = &kubeFunctionRunner{}
+
+func (kr *kubeFunctionRunner) Run(r io.Reader, w io.Writer) error {
+	// Wait until the pod is Running
+	wait.PollImmediate(100*time.Millisecond, 60*time.Second, func() (done bool, err error) {
+		err = kr.cl.Get(kr.ctx, client.ObjectKeyFromObject(kr.pod), kr.pod)
+		if err != nil {
+			return false, err
+		}
+		return kr.pod.Status.Phase == "Running", nil
+	})
+
+	address := kr.pod.Status.PodIP
+	if address == "" {
+		return fmt.Errorf("pod did not have podIP")
+	}
+	address += ":9446"
+
+	klog.Infof("dialing grpc function runner %q", address)
+
+	// TODO: pool connections
+	cc, err := grpc.Dial(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
```

Comment:
will this work? in what context is the port number getting resolved?
the target pod runs in a different namespace so we'll need some way to reach it, and I suspect the port number won't be enough, especially not if it is a constant.

Reply:
do we have guarantee that pod is listening on the gRPC port when the k8s signals the pod as "Running"?

I noticed that there is possible delay between the two events (but that could be because I was also spinning up Service resource)

Reply:
Added a readiness probe and check the ready condition.